design/httpproxy-conflict.md

```diff
@@ -0,0 +1,118 @@
+# HTTPProxy Conflict Resolution Design
+
+Status: Draft
+
+## Abstract
+
+Configuration errors in HTTPProxy resources can result in objects being marked as invalid and consequently will not serve traffic.
+These errors can come from invalid route configuration, missing service references, or invalid includes to other resources. 
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress that reduces annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to users via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, only the first is shown to the user.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes or even the entire virtual host. 
+
+### Personas
+
+There are two main personas that interact with this conflict design document:
+
+- User: A request to an application, routing through Envoy looking to access whatever resource an application running in Kubernetes should serve. This user is not aware of how the application is deployed or that it uses Kubernetes or Contour.
+- ResourceOwner: Someone to manages an HTTPProxy resource inside the Kubernetes cluster configuring the Route/Service combination to allow users of the application to access.  
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+
+## Non Goals
+- Change how Contour processes objects today from informers set against the Kubernetes API
+
+## High-Level Design
+Contour will set the error or warning condition for the when an error is encountered, but do its best to still serve valid configurations.
+
+It's difficult to find errors and block them from breaking valid configurations since Contour currently processes objects after they are committed to the API server.
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`)
+If Contour handles a configuration error by programming Envoy to return an error response instead of proxying the request through the route, the error will be published as an `Error` status.
+
+In general, Contour will set an object to be status `Error` if the request response to the `User` is changed from what is configured.
```

Comment:
I initially liked this way of describing it, but I think it's better to be much more specific about the exact behavior that an error indicates, versus a warning. That is, that an `Error` indicates that the object will not be processed by Contour - no part of it will end up as config. This is the behavior we have now. A Warning indicates that the object has a problem, but it's not significant enough for Contour not to be able to accept the object, so at least parts of the HTTPProxy should end up in Envoy.

This is not one hundred percent correct - I think there are edge cases, but I think it's closer to the main intent of what we're trying to do.

Given that, I'd suggest moving something like this to much earlier in the document, probably we should draw the distinction as part of the Background section. Something like this:

"When processing HTTPProxy objects, it turns out some problems should not stop the rest of the object from being processed. To distinguish between fatal-for-processing and not, we propose calling the former `Errors`, and the latter `Warnings`."

Then we've got the description of what an error and a warning are in this context out of the way early, and we can refer to them throughout this document, confident we're all on the same page.

Reply:
> Error indicates that the object will not be processed by Contour

This isn't my intent and get's to what @skriss was asking for. I'll update the doc with my thoughts, but an "Error indicates that **parts** of the ojbect  will not be processed by Contour, but other parts still may be". The goal of this design is to still allow portions of the object to be configured while others aren't. This means the object could have an "Error" status, but still serve traffic (which does sound weird).

Reply:
added suggested text